FIX degree in polynomial kernel should be a floating number

NolantheNerd opened the pull request:
#### Reference Issues/PRs

None - this fix is not associated with an issue.

#### What does this implement/fix? Explain your changes.

This fix only changes the docs.

The SpectralClustering class' `degree` parameter is constrained to positive integers, but the SpectralClustering class' docstring specifies that `degree` can be a float.

I've updated the docstring to require that `degree` be passed as an int.

Commit:
Fixes Degree Type Discrepency in Docstring

The SpectralClustering class' parameter `degree` is constrained to
positive integers, but the docstring specifies that `degree` can be a
float.

I've updated the docstring to require that `degree` be passed as an int.

glemaitre commented:
I think the documentation is correct but the code is wrong.

The degree will be used by a the polynomial kernel:

https://github.com/scikit-learn/scikit-learn/blob/93fa00c13c17bd28cfc95a1f6f3fd15ba02d8d4c/sklearn/metrics/pairwise.py#L1355-L1369

and this is constrained to be a float.

@jeremiedbb I see that I missed this thing in the following review: https://github.com/scikit-learn/scikit-learn/pull/24035/

We also tolerate an integer at 0 that will be excluded by the kernel itself. I assume that we could fore all the degrees parameter to `"degree": [Interval(Real, 0, None, closed="left")]`

@jeremiedbb WDYT?

glemaitre commented:
Interacting IRL with @jeremiedbb, we agreed that this is a regression and that we should use a `Real` number instead of `Integral` her. @NolantheNerd Do you want to contribute this change?

Commit:
Changed Degree Type Check for SpectralClustering

Changed degree type requirement from Integral to Real in
SpectralClustering class. This now matches both the SpectralClustering
docstring and the expected behaviour of degree, given that it is passed
to the polynomial kernel.

NolantheNerd committed:
Merge branch 'main' into docs/spectralclustering-degree-type

NolantheNerd commented:
Done! Ready for review!

glemaitre commented:
Can you also update:

- the degree parameter from `KernelPCA` to be a real as well
- the degree parameter from `KerelRidge` to be a real as well

In addition, we will need an entry in the changelog to acknowledge the regression.

Commit:
Merge remote-tracking branch 'upstream/main' into docs/spectralclustering-degree-type

Commit:
Merge branch 'docs/spectralclustering-degree-type' of github.com:LoKnow/scikit-learn into docs/spectralclustering-degree-type

Commit:
Degree Parameters Consistently Changed to Floats

In the SpectralClustering, KernelPCA and KernelRidge constructors, the
degree parameter was modified to allow float values, not just integral
values. The respective class' have had their docstrings updated to
reflect this change.

NolantheNerd commented:
Changes complete! Now for `cluster.SpectralClustering`,`kernel_ridge.KernelRidge` and `decomposition.KernelPCA`:

- The docstring reflects that `degree` is a float.
- The default value has been changed from `3` -> `3.0` in the docstring to further clarify that floats are permitted.
- The default value in the constructor has been changed from `3` -> `3.0`.

There doesn't seem to be a consistent pattern for integer/float disambiguation in the codebase, for example, in the KernelPCA docstring, both `1` and `1.0` are used for float values. @glemaitre Is there a "right" way to do it?
https://github.com/scikit-learn/scikit-learn/blob/4ca01961969a0c9e1c7c48410e0976bb04a92703/sklearn/decomposition/_kernel_pca.py#L62-L73

glemaitre commented:
> There doesn't seem to be a consistent pattern for integer/float disambiguation in the codebase

In this case both integral an float have the same behaviour so there is no need for disambiguation.

Sometimes we have float and integral that does not behave the same but usually the range of the accepted values will be different and the parameter validation will be here to help.

glemaitre commented on sklearn/decomposition/_kernel_pca.py:
```diff
@@ -278,7 +278,7 @@ def __init__(
         kernel="linear",
         gamma=None,
-        degree=3,
+        degree=3.0,
```
```suggestion
        degree=3,
```

glemaitre commented on sklearn/decomposition/_kernel_pca.py:
```diff
@@ -56,7 +56,7 @@ class KernelPCA(ClassNamePrefixFeaturesOutMixin, TransformerMixin, BaseEstimator
         kernels. If ``gamma`` is ``None``, then it is set to ``1/n_features``.
 
-    degree : int, default=3
+    degree : float, default=3.0
```
I think we can let 3 here because both 3 and 3.0 behave the same.

```suggestion
    degree : float, default=3
```

glemaitre commented on sklearn/kernel_ridge.py:
```diff
@@ -149,7 +149,7 @@ def __init__(
         kernel="linear",
         gamma=None,
-        degree=3,
+        degree=3.0,
```
```suggestion
        degree=3,
```

glemaitre commented on sklearn/kernel_ridge.py:
```diff
@@ -69,7 +69,7 @@ class KernelRidge(MultiOutputMixin, RegressorMixin, BaseEstimator):
         Ignored by other kernels.
 
-    degree : int, default=3
+    degree : float, default=3.0
```
```suggestion
    degree : float, default=3
```

glemaitre commented on sklearn/cluster/_spectral.py:
```diff
@@ -481,7 +481,7 @@ class SpectralClustering(ClusterMixin, BaseEstimator):
            Added new labeling method 'cluster_qr'.
 
-    degree : float, default=3
+    degree : float, default=3.0
```
```suggestion
    degree : float, default=3
```

glemaitre commented on sklearn/cluster/_spectral.py:
```diff
@@ -629,7 +629,7 @@ def __init__(
         eigen_tol="auto",
         assign_labels="kmeans",
-        degree=3,
+        degree=3.0,
```
```suggestion
        degree=3,
```

glemaitre reviewed (approved):
I only think that we can keep 3 instead of 3.0 since this is already a valid Real.

Otherwise LGTM. Thanks @NolantheNerd

NolantheNerd committed:
Update sklearn/decomposition/_kernel_pca.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

NolantheNerd committed:
Update sklearn/decomposition/_kernel_pca.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

NolantheNerd committed:
Update sklearn/kernel_ridge.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

NolantheNerd committed:
Update sklearn/kernel_ridge.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

NolantheNerd committed:
Update sklearn/cluster/_spectral.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

NolantheNerd committed:
Update sklearn/cluster/_spectral.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

OmarManzoor reviewed (approved):
LGTM. Thanks @NolantheNerd

NolantheNerd committed:
Merge branch 'main' into docs/spectralclustering-degree-type

NolantheNerd commented:
@OmarManzoor looks like there was a conflict in the news file. I've fixed the conflict.
